doc/source/data/data-internals.rst

```diff
@@ -11,79 +11,164 @@ For a gentler introduction to Ray Data, see :ref:`Quickstart <data_quickstart>`.
-   :align: center
-..
-  https://docs.google.com/drawings/d/12STHGV0meGWfdWyBlJMUgw7a-JcFPu9BwSOn5BjRw9k/edit
+Operators
+~~~~~~~~~
+
+There are two types of operators: *logical operators* and *physical operators*. Logical 
+operators are stateless objects that describe “what” to do. Physical operators are 
+stateful objects that describe “how” to do it. An example of a logical operator is 
+``ReadOp``, and an example of a physical operator is ``TaskPoolMapOperator``.
+
+Plans
+~~~~~
+
+A *logical plan* is a series of logical operators, and a *physical plan* is a series of 
+physical operators. When you call APIs like :func:`ray.data.read_images` and 
+:meth:`ray.data.Dataset.map_batches`, Ray Data produces a logical plan. When execution 
+starts, the planner generates a corresponding physical plan. 
+
+The planner
+~~~~~~~~~~~
+
+The Ray Data planner translates logical operators to one or more physical operators. For 
+example, the planner translates the ``ReadOp`` logical operator into two physical 
+operators: an ``InputDataBuffer`` and ``TaskPoolMapOperator``. Whereas the ``ReadOp``
+logical operator only describes the input data, the ``TaskPoolMapOperator`` physical 
+operator actually launches tasks to read the data.
+
+Plan optimization
+~~~~~~~~~~~~~~~~~
+
+Ray Data applies optimizations to both logical and physical plans. For example, the 
+``OperatorFusionRule`` combines a chain of physical map operators into a single map 
+operator. This prevents unnecessary serialization between map operators.
+
+Types of physical operators
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Physical operators take in a stream of block references and output another stream of 
+block references. Some physical operators launch Ray Tasks and Actors to transform  
+the blocks, and others only manipulate the references.
 
-For the preceding example, each read task generates one output block. In the case where the read input file
-is large, a single read task may output multiple blocks. For more information on transforming data, see
-:ref:`Transforming data <transforming_data>`.
+``MapOperator`` is the most common operator. All read, transform, and write operations 
+are implemented with it. To process data, ``MapOperator`` implementations use either Ray 
+Tasks or Ray Actors.
 
-Shuffling data
---------------
+Non-map operators include ``OutputSplitter`` and ``LimitOperator``. These two operators 
+manipulate references to data, but don’t launch tasks or modify the underlying data. 
 
-When you call :meth:`~ray.data.Dataset.random_shuffle`,
-:meth:`~ray.data.Dataset.sort`, or :meth:`~ray.data.Dataset.groupby`, Ray Data shuffles
-blocks in a map-reduce style: map tasks partition blocks by value and then reduce tasks
-merge co-partitioned blocks.
+Execution
+---------
+
+The executor
+~~~~~~~~~~~~
+
+The *executor* schedules tasks and moves data between physical operators.
+
+The executor and operators are located on the driver process, except for training jobs, 
```

Comment:
ditto. worth mentioning that the executor and operators are created on the process where dataset execution starts. for batch inference jobs, the process is usually the driver. while users can also pass the Dataset object to any Ray actors or tasks and trigger execution there. for training jobs, the executor runs on a special actor called SplitCoordinator which is also responsible for handling `streaming_split`.